File: users/server/internal/mail/smtp.go
package mail

import (
	"fmt"
	"net/smtp"
	"strings"
)

type Config struct {
	Host     string
	Port     string
	User     string
	Password string
	From     string
}

type Sender struct {
	cfg Config
}

func NewSender(cfg Config) *Sender {
	return &Sender{cfg: cfg}
}

func (s *Sender) Configured() bool {
	return s.cfg.Host != "" && s.cfg.User != "" && s.cfg.Password != "" && s.cfg.From != ""
}

func (s *Sender) Send(to []string, subject, body string) error {
	if !s.Configured() {
		return fmt.Errorf("mail not configured")
	}
	addr := fmt.Sprintf("%s:%s", s.cfg.Host, s.cfg.Port)
	auth := smtp.PlainAuth("", s.cfg.User, s.cfg.Password, s.cfg.Host)
	msg := []byte(fmt.Sprintf("From: %s\r\nTo: %s\r\nSubject: %s\r\nMIME-Version: 1.0\r\nContent-Type: text/plain; charset=utf-8\r\n\r\n%s\r\n",
		s.cfg.From, strings.Join(to, ", "), subject, body))
	return smtp.SendMail(addr, auth, s.cfg.From, to, msg)
}

File: users/server/src/mail/smtp.ts
import nodemailer from "nodemailer";
import type { Config } from "../config.ts";
import { mailConfigured } from "../config.ts";

export function createMailer(cfg: Config) {
  const configured = mailConfigured(cfg);
  const transporter = configured
    ? nodemailer.createTransport({
        host: cfg.smtpHost,
        port: Number(cfg.smtpPort),
        auth: { user: cfg.smtpUser, pass: cfg.smtpPassword },
      })
    : null;

  return {
    configured: () => configured,
    send: async (to: string[], subject: string, body: string) => {
      if (!transporter || !configured) throw new Error("mail not configured");
      await transporter.sendMail({
        from: cfg.mailFrom,
        to,
        subject,
        text: body,
      });
    },
  };
}

export type Mailer = ReturnType<typeof createMailer>;
